launcher/launcher.rb

```diff
@@ -0,0 +1,147 @@
+    Process.kill 'KILL', File.read(PID).to_i if File.exist?(PID)
+  rescue
+  end
+end
+
+
+def desktop_icon_name_finder (app)
+  # find out the info, name and icon of the command
+    
+  @desktop = `find / -name #{app}.desktop 2> /dev/null | sed -e '/stateful_partition/d' | tail -n1`.tr("\n", '')
+    
+  # abort if a desktop file not found for the command
+  if @desktop == '' then
+    puts "Cannot find the .desktop file for #{app}, using default Chromebrew icon... 
+(If you want to use customize icon, put the icon file to #{CREW_PREFIX}/lib/pwa/ and name it to '#{app}.png')".yellow
+    FileUtils.mkdir_p PWA_PREFIX
+    system('curl', '-L#', 'https://github.com/supechicken/chromebrew/raw/master/images/brew546.png', '-o', PWA_PREFIX + '/brew.png')
+    @icon = PWA_PREFIX + '/brew.png'
+    @icon = PWA_PREFIX + "/#{app}.png" if File.exist?(PWA_PREFIX + "/#{app}.png")
+    @name = app.capitalize
+  else
+    @icon = File.readlines(@desktop).select { |line| line =~ /^Icon/ }[0].partition('=')[2].tr("\n", '')
+    @icon = `find #{PATH} -iname #{@icon}.png -o -iname #{@icon}.xpm 2> /dev/null | sed -e '/stateful_partition/d' |\
+             sed -e 's:/usr/local/share/icons/hicolor/::g' | sort -n | sed -e 's:^:/usr/local/share/icons/hicolor/:g' |\
+             tail -n1`.tr("\n", '')
+  
+
+    @name = File.readlines(@desktop).select { |line| line =~ /^Name/ }[0].partition('=')[2].tr("\n", '')
+    @info = File.readlines(@desktop).select { |line| line =~ /^Comment/ }[0].partition('=')[2].tr("\n", '')
+  end
+  @info = `crew search -d #{app} | head -1` unless @info
+end
+
+def new (app)
+  # abort when command not found
+  abort "#{app}: command not found".lightred unless system("which #{app} > /dev/null 2>1")
+
+  desktop_icon_name_finder(app)
+  
+  @working_on_dir = "#{Dir.home}/MyFiles/#{app}.launcher/"
+  FileUtils.mkdir_p(@working_on_dir)
+  
+  system('convert', @icon, '-resize', '1024x1024', @working_on_dir + 'icon.png')
+    
+  File.write @working_on_dir + 'manifest.json', JSON.gsub('NAME', @name).gsub('INFO', @info)
+  File.write @working_on_dir + 'background.js', JS.gsub('CMD', app)
+  
+  puts POST_INSTALL.gsub('<CMD>', app).lightblue
+end
+
+def start_server
+    
+  stop_server
+    
+  system('startsommelier > /dev/null') or abort "Failed to start graphical environment: sommelier exited with status #{$?.exitstatus.to_s}".lightred
+  ENV['DISPLAY'] = ':0'
+  ENV['GDK_BACKEND'] = 'x11'
+  Process.fork do
+    # write current fork process pid to a file
+    File.write(PID, Process.pid)
+    EM.run {
+      EM::WebSocket.run(:host => "0.0.0.0", :port => 25500) do |ws|
+        ws.onmessage { |message|
+          File.open(LOG, 'w') { |file|
+            $stdout.sync = true
+            $stdout.reopen(file)
+            $stderr.sync = true
+            $stderr.reopen(file)
+          }
+          system(message + '&')
+        }
+      end
+    }
+  end
+end
+    
+case ARGV[0]
+when 'start'
+  start_server
+when 'new'
+  new ARGV[1]
```

Comment:
Will this return a clear message if `ARGV[1]` is missing?  Would it make sense to echo the usage text instead?

Reply:
Added